components/banners.js

```diff
@@ -0,0 +1,82 @@
+import Alert from 'react-bootstrap/Alert'
+import styles from './banners.module.css'
+import { useEffect, useState } from 'react'
+import { useMe } from '../components/me'
+import { useMutation } from '@apollo/client'
+import { WELCOME_BANNER_MUTATION } from '../fragments/users'
+import { useToast } from '../components/toast'
+
+export default function NewVisitorBanner () {
+  const me = useMe()
+  const toaster = useToast()
+  const [state, setState] = useState(null)
+  const handleClose = async () => {
+    const obj = { date: state?.date || new Date().toJSON(), hidden: true }
+    const str = JSON.stringify(obj)
+    window.localStorage.setItem('newVisitorInfo', str)
+    setState(obj)
+    if (me) {
+      let error
+      try {
+        ({ error } = await hideWelcomeBanner({ variables: { } }))
+      } catch (e) {
+        error = e
+      }
+      if (error) toaster.danger(error.toString())
+    }
+  }
+  const [hideWelcomeBanner] = useMutation(WELCOME_BANNER_MUTATION, {
+    update (cache) {
+      cache.modify({
+        id: `User:${me.id}`,
+        fields: {
+          hideWelcomeBanner () {
+            return true
+          }
+        }
+      })
+    }
+  })
+  useEffect(() => {
+    let str = window.localStorage.getItem('newVisitorInfo')
+    let obj = JSON.parse(str)
+    if (!str) {
+      obj = { date: new Date().toJSON(), hidden: false }
+    }
+    if (me?.hideWelcomeBanner) {
+      obj.hidden = true
+    }
+    str = JSON.stringify(obj)
+    window.localStorage.setItem('newVisitorInfo', str)
+    setState(obj)
+  }, [me?.hideWelcomeBanner])
+
+  if ((me && !me.hideWelcomeBanner) || (!me && state && !state.hidden)) {
```

Comment:
You can use [optional chaining](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Optional_chaining):

```suggestion
  if (!me?.hideWelcomeBanner || !state?.hidden) {
```

Reply:
BTW, don't be confused if I propose code changes to code that I already suggested to change in other ways which would make this change unnecessary, lol

Just want to let you know how this line of code could be better written _if_ there is no reason to not include it at all :)

Reply:
My very explicit logic was actually intentional for clarity, and there is a subtle difference between `me && !me.x` vs. `!me?.x` that makes the shorthand not work here. To be clear, in this case, `!me?.hideWelcomeBanner` evaluates to true when not logged in, which causes the banner to appear regardless of the other flag. But with all the other reductions, at least the right-hand side of the condition has now been simplified.

Reply:
interesting, you are right. Nice catch!